File: clients/unity-app/Assets/Scripts/Labs/Lab02.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using Colyseus;
using Colyseus.Predict;
using ColyseusSchema = Colyseus.Schema.Schema;
using UnityEngine;
using Lab = PredictProbe.LabSchema;

namespace Playground
{
    /// <summary>
    /// Lab 02 — Clocks &amp; Timelines. There is no clock "API call" to make: the
    /// moment the server declares defineInput(), every input round-trip carries a
    /// TIMED prefix and the SDK maintains room.Clock. This lab just reads it —
    /// Now / ServerNow / RenderNow / Rtt / SmoothedRtt / Jitter / LastServerTime.
    /// Port of labs/02-clocks/.
    /// </summary>
    public class Lab02 : LabBase<Lab.BotsState>
    {
        private const double StripSpanMs = 3000;

        public override string Id => "02-clocks";
        public override int Num => 2;
        public override string Title => "Clocks & Timelines";
        public override string Blurb => "serverNow / renderNow / RTT / jitter.";

        private Lab.MoveInput _cmd;
        private InputHandle _input;
        private Pacer _pacer;
        private string _sid;
        private Lab.Bot _bot;

        private readonly List<double> _arrivals = new List<double>();
        private double _lastPatchStamp, _sparkGate;
        private readonly Spark _rtt = new Spark(), _age = new Spark(), _slew = new Spark();


        public override async Task<bool> Mount(App app)
        {
            Room = await Shell.JoinLab<Lab.BotsState>(app, "lab-bots",
                r => r.State.bots != null && r.State.bots.ContainsKey("bot1"));
            _sid = Room.SessionId;
            if (Room.State.bots == null || !Room.State.bots.TryGetValue("bot1", out _bot)) return false;
            _cmd = new Lab.MoveInput();
            // Inputs feed the clock: one send per fixed tick = one RTT/offset sample.
            _input = Room.Input(_cmd);
            _pacer = new Pacer(1000.0 / Sim.TickHz);
            return true;
        }

        public override void Frame(App app, double now, double dtMs)
        {
            int steps = _pacer.Steps(now);
            for (int i = 0; i < steps; i++)
            {
                _cmd.moveX = (sbyte)Kb.MoveX();
                _cmd.moveY = (sbyte)Kb.MoveY();
                _input.Send();
            }

            // A patch arrival = the server stamp changing.
            double stamp = Room.Clock.LastServerTime();
            if (stamp != _lastPatchStamp)
            {
                _lastPatchStamp = stamp;
                _arrivals.Add(now);
                if (_arrivals.Count > 128) _arrivals.RemoveAt(0);
            }

            _age.Push(System.Math.Max(0, Room.Clock.ServerNow() - stamp));
            _sparkGate += dtMs;
            if (_sparkGate >= 200)
            {
                _sparkGate = 0;
                _rtt.Push(Room.Clock.Rtt());
                _slew.Push(System.Math.Abs(Room.Clock.ServerNow() - Room.Clock.RenderNow()));
            }
        }

        public override void Render(App app)
        {
            var v = app.View;
            Room.State.players.ForEach((key, p) =>
                Draw.Square(v, p.x, p.y, Sim.PlayerHalf, Palette.Hue(p.hue, key == _sid ? 0.9f : 0.4f)));

            Draw.CircleOutline(v, _bot.x, _bot.y, Sim.BotRadius, Palette.Text, dashed: true);
            Draw.Label(v, _bot.x, _bot.y, "raw snapshots (patch rate)",
                Palette.A(Palette.Text, 0.6f), 10, -v.S(Sim.BotRadius) - 18);

            // Patch-arrival strip along the bottom of the stage.
            float x0 = app.Stage.x + 24, x1 = app.Stage.xMax - 20, y = app.Stage.yMax - 34;
            if (x1 - x0 > 200)
            {
                Draw.Line(new Vector2(x0, y), new Vector2(x1, y), Palette.A(Palette.TextDim, 0.3f));
                double now = RoomClock.GetNow();
                foreach (double t in _arrivals)
                {
                    double age = now - t;
                    if (age > StripSpanMs) continue;
                    float px = x1 - (float)(age / StripSpanMs) * (x1 - x0);
                    Draw.Rect(new Rect(px - 2, y - 2, 4, 4), Palette.Good);
                }
                var s = new GUIStyle(GUI.skin.label) { fontSize = 10, normal = { textColor = Palette.A(Palette.Text, 0.55f) } };
                GUI.Label(new Rect(x0, y - 18, 420, 14), "patch arrivals (local clock, last 3 s) ->", s);
            }

            var h = app.Hud;
            h.Section("TELEMETRY");
            h.Spark(_rtt, "rtt", $"{_rtt.Last:F0} ms", Palette.Blue);
            h.Spark(_age, "patch age (sawtooth)", $"{_age.Last:F0} ms", Palette.Good);
            h.Spark(_slew, "|serverNow - renderNow| (slew)", $"{_slew.Last:F1} ms", Palette.Warn);
            h.Row("serverNow", $"{Room.Clock.ServerNow():F0} ms", Palette.Text);
            h.Row("renderNow (slewed)", $"{Room.Clock.RenderNow():F0} ms", Palette.Text);
            h.Row("jitter (interarrival)", $"{Room.Clock.Jitter():F1} ms", Palette.Text);
            h.Row("smoothed rtt", $"{Room.Clock.SmoothedRtt():F0} ms", Palette.Text);
            h.Row("patch interval", $"{Room.Clock.PatchInterval():F0} ms", Palette.Text);

            h.Section("CONTROLS");
            h.Key("WASD", "drive");
            h.Note("Nothing to configure here — yank the latency preset with L and watch every " +
                   "readout respond. The offset re-converges; the slew spike decays over ~250 ms. " +
                   "The dot row marks each PATCH ARRIVAL on the local clock: add jitter and watch " +
                   "the spacing get ragged.");
        }

        public override void Unmount() { _arrivals.Clear(); }

        public override void OnReconnect()
        {
            _pacer.Reset();
            _arrivals.Clear();
            _lastPatchStamp = 0;
            Room.State.bots.TryGetValue("bot1", out _bot);
        }
    }
}

File: clients/defold-app/playground/labs/lab02.lua
--
-- Lab 02 — Clocks & Timelines.
--
-- There is no clock "API call" to make: the moment the server declares
-- defineInput(), every input round-trip carries a TIMED prefix and the SDK
-- maintains room.clock. This lab just reads it — now / server_now / render_now /
-- rtt / smoothed_rtt / jitter / last_server_time.
--
-- Port of src/client/labs/02-clocks/.
--
local sim = require 'playground.sim'
local app = require 'playground.app'

local kb = app.kb
local STRIP_SPAN_MS = 3000

local Lab = {
  id = "02-clocks",
  num = 2,
  title = "Clocks & Timelines",
  blurb = "server_now / render_now / RTT / jitter.",
  room_name = "lab-bots",
}
Lab.__index = Lab

function Lab.new()
  return setmetatable({ arrivals = {}, last_patch_stamp = 0 }, Lab)
end

function Lab:ready(room)
  return room.state.bots ~= nil and room.state.bots["bot1"] ~= nil
end

function Lab:mount(context, room)
  self.room = room
  self.sid = room.session_id
  self.bot = room.state.bots["bot1"]
  if self.bot == nil then return false end
  -- Inputs feed the clock: one send per fixed tick = one RTT/offset sample.
  self.input = room:input()
  self.cmd = self.input.data
  self.pacer = app.Pacer.new(1000.0 / sim.TICK_HZ)
  return true
end

function Lab:frame(context, now, dt_ms)
  local steps = self.pacer:steps(now)
  for _ = 1, steps do
    self.cmd.moveX = kb.move_x()
    self.cmd.moveY = kb.move_y()
    self.input:send()
  end

  -- A patch arrival = the server stamp changing.
  local stamp = self.room.clock:last_server_time()
  if stamp ~= self.last_patch_stamp then
    self.last_patch_stamp = stamp
    table.insert(self.arrivals, now)
    if #self.arrivals > 128 then table.remove(self.arrivals, 1) end
  end
end

function Lab:render(gfx)
  local clock = self.room.clock

  self.room.state.players:each(function(p, key)
    gfx.square(p.x, p.y, sim.PLAYER_HALF, gfx.hue(p.hue, key == self.sid and 0.9 or 0.4))
  end)

  gfx.circle_outline(self.bot.x, self.bot.y, sim.BOT_RADIUS, gfx.PALETTE.text, true)
  gfx.label(self.bot.x, self.bot.y, "raw snapshots (patch rate)",
    gfx.a(gfx.PALETTE.text, 0.6), 10, -18)

  gfx.hud_section("TELEMETRY")
  gfx.hud_row("rtt", string.format("%.0f ms", clock:rtt()), gfx.PALETTE.blue)
  gfx.hud_row("smoothed rtt", string.format("%.0f ms", clock:smoothed_rtt()), gfx.PALETTE.text)
  gfx.hud_row("patch age (sawtooth)",
    string.format("%.0f ms", math.max(0, clock:server_now() - clock:last_server_time())),
    gfx.PALETTE.good)
  gfx.hud_row("|server_now - render_now| (slew)",
    string.format("%.1f ms", math.abs(clock:server_now() - clock:render_now())), gfx.PALETTE.warn)
  gfx.hud_row("server_now", string.format("%.0f ms", clock:server_now()), gfx.PALETTE.text)
  gfx.hud_row("render_now (slewed)", string.format("%.0f ms", clock:render_now()), gfx.PALETTE.text)
  gfx.hud_row("jitter (interarrival)", string.format("%.1f ms", clock:jitter()), gfx.PALETTE.text)
  gfx.hud_row("patch interval", string.format("%.0f ms", clock:patch_interval()), gfx.PALETTE.text)
  gfx.hud_row("patch arrivals buffered", tostring(#self.arrivals), gfx.PALETTE.text)

  gfx.hud_section("CONTROLS")
  gfx.hud_key("WASD", "drive")
  gfx.hud_note("Nothing to configure here — yank the latency preset with L and " ..
    "watch every readout respond. The offset re-converges; the slew spike decays " ..
    "over ~250 ms.")
end

function Lab:unmount() self.arrivals = {} end

function Lab:on_reconnect()
  self.pacer:reset()
  self.arrivals = {}
  self.last_patch_stamp = 0
  self.bot = self.room.state.bots["bot1"]
end

return Lab
